### ISSUE TITLE
Alt key press, keyboard shortcut and support for Emacs keyboard command.

### ISSUE THREAD

**Issue by jnsebgosselin** [spyder-ide/spyder#7981]
## Problem Description

As can be seen in the card below, a lot of the shortcuts defined in Emacs consist of the `Alt` key and another key. Note that the `Alt` modifier key corresponds to the symbol `M` in the table.

However, on Windows, the focus is automatically given to the mainwindow's menubar when the `Alt` key is held down. If one pressed another key, like `F` for example, while the `Alt` key is down, it expands the menu starting with that letter. So in Spyder, if the user does `Alt+F`, it resuts in the `File` menu being expanded and focused.

So... this will obviously cause problems for the support of Emacs keybindings. I can think of two options:

- We use something else to replace the `Alt + <another key>` shortcuts.
- We override the default behavior in Windows and prevent the menubar from grabbing focus after Alt is pressed.

So what do you guys think? @ccordoba12 @jitseniesen @CAM-Gerlach 

https://www.gnu.org/software/emacs/refcards/pdf/refcard.pdf
![image](https://user-images.githubusercontent.com/10170372/46351793-3d0b2e80-c626-11e8-8257-23afdb4863d8.png)

Resources:
https://stackoverflow.com/questions/37020992/qt-prevent-menubar-from-grabbing-focus-after-alt-pressed-on-windows

**Comment by ccordoba12** [comment]
> We override the default behavior in Windows and prevent the menubar from grabbing focus after Alt is pressed.

This is what we need to do because Emacs shortcuts are ingrained in muscle memory for its long time users.

**Comment by CAM-Gerlach** [comment]
> >  We override the default behavior in Windows and prevent the menubar from grabbing focus after Alt is pressed.

> This is what we need to do because Emacs shortcuts are ingrained in muscle memory for its long time users.

The problem is that Alt-<F/E/V/etc>-<Scroll/Key>l-Enter is ingrained in muscle memory for many longtime Windows users for quickly accessing menu commands, and I frequently use it myself. I don't recall a single Windows-native or cross-platform application I've ever used that's broken this convention, and as generally more "power users", people using Spyder are more likely than average to employ them. 

Furthermore, I believe (although I'm not 100% certain) that Alt-Spacebar relies on the behavior as well, which allows quick access to the minimize, maximize, restore, etc. window commands—and most crucially, Alt-Spacebar, c is the standard window close command (equivalent to Cmd-Q on Windows) that's drilled into my muscle memory and I've never, ever seen an application override it.

Therefore, given this is basic, nearly universal OS-standard behavior that users expect of any Windows application, I don't see how we can justify overriding it on that platform by default to cater to a minority of longtime users of one particular editor native to a different platform. However, I see no problem with only doing so if an Emacs shortcut preset is enabled (perhaps with a warning dialog or other admonition so users aren't surprised when the functionality stops working), or offering it as a checkbox option in the keyboard shortcut preferences, off by default on Windows. We should ensure Alt-Spacebar still works, since that triggers OS-level functionality (equivalent to blocking Alt-Tab or Ctrl-Alt-Delete) and leaves the user without the means to easily close (or minimize, etc) Spyder itself on Windows.

**Comment by jnsebgosselin** [comment]
I am not comfortable with overriding the default behavior on Windows also... but I don't know how important support for Emacs is...

Personally, I don't care much about Emacs. I'm more focused about fixing everything that is not working and make the keyboard shortcuts as much configurable as possible.

So bottom line, supporting the entire set of Emacs commands on Windows is not going to be an easy task...

**Comment by CAM-Gerlach** [comment]
> but I don't know how important support for Emacs is...

If it really is that important, an off-by default option configurable by a checkbox or by explicitly enabling the emacs shortcuts/emacs mod would be the best of both worlds between the advantages of both options you present, without either changing the default Emacs shortcuts or disabling OS behavior by default, unless the user explicitly preferred it.

**Comment by ccordoba12** [comment]
Yeah, the Alt behavior for Emacs should be activated only if the Emacs shortcuts set is selected.

**Comment by jnsebgosselin** [comment]
> Yeah, the Alt behavior for Emacs should be activated only if the Emacs shortcuts set is selected.

This solution doesn't take into account custom shortcuts that are defined by the users though...

I think we need to add an option configurable by a checkbox instead in the `Keyboard Shortcut` preferences panel.

**Comment by CAM-Gerlach** [comment]
> I think we need to add an option configurable by a checkbox instead in the Keyboard Shortcut preferences panel.

Agreed, that's the best option—best be explicit about it.

**Comment by ccordoba12** [comment]
Ok, sure, that's a good option too.
